### PR TITLE
Add Book and BookShelf models

### PR DESCRIPTION
**Pull request by takuyakanbr**
* Add Book and its components
* Add UniqueBookList
* Add BookShelf
* Add unit tests for new classes

### CODE REVIEW

File: src/main/java/seedu/address/model/BookShelf.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        requireNonNull(newData);
+
+        try {
+            setBooks(newData.getBookList());
+        } catch (DuplicateBookException e) {
+            throw new AssertionError("BookShelf should not have duplicate books");
+        }
+    }
+
+    //// person-level operations
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/BookShelf.java]: book

File: src/main/java/seedu/address/model/ModelManager.java [hunk 2]
```diff
@@ -12,35 +12,64 @@
...
+     */
+    public ModelManager(ReadOnlyBookShelf bookShelf, UserPrefs userPrefs) {
+        super();
+        requireAllNonNull(bookShelf, userPrefs);
+
+        logger.fine("Initializing with book shelf: " + bookShelf + " and user prefs " + userPrefs);
+
+        this.bookShelf = new BookShelf(bookShelf);
+        this.filteredBooks = new FilteredList<>(this.bookShelf.getBookList());
+        this.addressBook = new AddressBook();
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/ModelManager.java]: remove

File: src/main/java/seedu/address/model/ModelManager.java [hunk 3]
```diff
@@ -49,6 +78,74 @@ public void resetData(ReadOnlyAddressBook newData) {
...
+
+    /** Raises an event to indicate the model has changed */
+    private void indicateBookShelfChanged() {
+        raise(new BookShelfChangedEvent(bookShelf));
+    }
+
+    @Override
+    public void deleteBook(Book target) throws BookNotFoundException {
+        bookShelf.removeBook(target);
+        indicateAddressBookChanged();
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/ModelManager.java]: indicateBookShelfChanged

File: src/main/java/seedu/address/model/ModelManager.java [hunk 4]
```diff
@@ -49,6 +78,74 @@ public void resetData(ReadOnlyAddressBook newData) {
...
+    }
+
+    @Override
+    public void addBook(Book book) throws DuplicateBookException {
+        bookShelf.addBook(book);
+        updateFilteredBookList(PREDICATE_SHOW_ALL_BOOKS);
+        indicateAddressBookChanged();
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/ModelManager.java]: indicateBookShelfChanged

File: src/main/java/seedu/address/model/ModelManager.java [hunk 5]
```diff
@@ -49,6 +78,74 @@ public void resetData(ReadOnlyAddressBook newData) {
...
+    }
+
+    @Override
+    public void updateBook(Book target, Book editedBook) throws BookNotFoundException, DuplicateBookException {
+        requireAllNonNull(target, editedBook);
+
+        bookShelf.updateBook(target, editedBook);
+        indicateAddressBookChanged();
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/ModelManager.java]: indicateBookShelfChanged

File: src/main/java/seedu/address/model/ModelManager.java [hunk 6]
```diff
@@ -49,6 +78,74 @@ public void resetData(ReadOnlyAddressBook newData) {
...
+
+        // instanceof handles nulls
+        if (!(obj instanceof ModelManager)) {
+            return false;
+        }
+
+        // state check
+        ModelManager other = (ModelManager) obj;
+        return addressBook.equals(other.addressBook) // TODO switch to using book shelf
+                && filteredPersons.equals(other.filteredPersons);
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/ModelManager.java]: change to bookshelf and filteredBooks

File: src/main/java/seedu/address/model/UniqueList.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+        assert CollectionUtil.elementsAreUnique(internalList);
+    }
+
+    /**
+     * Returns all items in this list as a Set.
+     * This set is mutable and change-insulated against the internal list.
+     */
+    public Set<T> toSet() {
+        assert CollectionUtil.elementsAreUnique(internalList);
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/UniqueList.java]: why assert here?

Comment (qiu-siqi) [reply src/main/java/seedu/address/model/UniqueList.java]: set is unique anyway

File: src/main/java/seedu/address/model/book/Author.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * Guarantees: immutable.
+ */
+public class Author {
+
+    public final String fullName;
+
+    /**
+     * Constructs a {@code Author}.
+     *
+     * @param fullName A valid author name.
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/Author.java]: not checked for validity

File: src/main/java/seedu/address/model/book/Book.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        if (other == this) {
+            return true;
+        }
+
+        if (!(other instanceof Book)) {
+            return false;
+        }
+
+        Book otherBook = (Book) other;
+        return otherBook.getTitle().equals(this.getTitle());
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/Book.java]: not the best idea... may need to check authors too

File: src/main/java/seedu/address/model/book/Book.java [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(title);
+    }
+
+    @Override
+    public String toString() {
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/Book.java]: you mean this looks legit? o_o

File: src/main/java/seedu/address/model/book/Category.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+ * Guarantees: immutable.
+ */
+public class Category {
+
+    public final String category;
+
+    /**
+     * Constructs a {@code Category}.
+     *
+     * @param category A valid category.
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/Category.java]: what does valid even mean here

File: src/main/java/seedu/address/model/book/Description.java [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+ * Guarantees: immutable.
+ */
+public class Description {
+
+    public final String description;
+
+    /**
+     * Constructs a {@code Description}.
+     *
+     * @param description A valid description.
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/Description.java]: you must love the word valid a lot

File: src/main/java/seedu/address/model/book/UniqueBookList.java [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+import javafx.collections.ObservableList;
+import seedu.address.model.book.exceptions.BookNotFoundException;
+import seedu.address.model.book.exceptions.DuplicateBookException;
+
+/**
+ * Represents a unique collection of books. Does not allow nulls.
+ *
+ * Supports a minimal set of list operations.
+ */
+public class UniqueBookList implements Iterable<Book> {
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/UniqueBookList.java]: Not extending UniqueList?

File: src/main/java/seedu/address/model/book/UniqueBookList.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            throw new BookNotFoundException();
+        }
+        return bookFoundAndDeleted;
+    }
+
+    public void setBooks(UniqueBookList replacement) {
+        this.internalList.setAll(replacement.internalList);
+    }
+
+    public void setBooks(List<Book> books) throws DuplicateBookException {
```
Comment (qiu-siqi) [inline src/main/java/seedu/address/model/book/UniqueBookList.java]: can replace with Collection<Book>

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (takuyakanbr) [reply src/main/java/seedu/address/model/ModelManager.java]: This is needed for now to keep tests from failing

**Pull request merged**